### PR TITLE
Added DeepSpeed distributed strategy and backend

### PR DESCRIPTION
**Pull request by tgaddair**
This PR introduces two ways to run DeepSpeed to train large models (like LLMs):

- DeepSpeed backend
- Ray backend + DeepSpeed strategy

**DeepSpeed backend:**

Config:

```
backend:
    type: deepspeed
    zero_optimization:
        stage: 3
```

CLI:

```
deepspeed --num_gpus 4 ludwig train --config config.yaml --dataset dataset.csv
```

**Ray backend + DeepSpeed strategy:**

Config:

```
backend:
    type: ray
    trainer:
        strategy:
            type: deepspeed
            zero_optimization:
                stage: 3
        num_workers: 4
        use_gpu: True
```

CLI:

```
ray exec ludwig train --config config.yaml --dataset dataset.csv
```

Note that using the DeepSpeed backend will use Pandas for data processing, and as such does not support larger-than-memory datasets. For large datasets, use the Ray backend with the DeepSpeed strategy.

TODO:

- [x] Fix saving and loading for sharded models
- [x] Fix Ray backend
- [ ] Fix batch size tuning for DeepSpeed backend
- [x] Fix batch size tuning with Ray backend (requires #2934).
- [x] Support DeepSpeed [optimizers](https://deepspeed.readthedocs.io/en/latest/optimizers.html).

Known issues / limitations:

- DeepSpeed cannot run the model in eval mode (see: https://github.com/microsoft/DeepSpeed/issues/3068). This means dropout, batch norm, etc. will still be enabled when computing validation / test metrics.

### CODE REVIEW

File: examples/fintuning/train_imdb.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    experiment_name="imdb_sentiment",
+    model_name="parallel_cnn",
+)
+
+# list contents of output directory
+print("contents of output directory:", output_directory)
+for item in os.listdir(output_directory):
+    print("\t", item)
+
+# batch prediction
```
Comment (geoffreyangus) [inline examples/fintuning/train_imdb.py:78]: Remove?

File: examples/fintuning/train_imdb.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+from ludwig.api import LudwigModel
+
+# from ludwig.datasets import imdb
+
+# clean out prior results
+shutil.rmtree("./results", ignore_errors=True)
+# model_dir = "/src/results/simple_experiment_simple_model/model"
+
+# Download and prepare the dataset
+# training_set, test_set, _ = titanic.load(split=True)
```
Comment (geoffreyangus) [inline examples/fintuning/train_imdb.py:16]: Remove?

File: examples/titanic/simple_model_training.py [hunk 3]
```diff
@@ -13,13 +13,15 @@
 import yaml
 
 from ludwig.api import LudwigModel
-from ludwig.datasets import titanic
+
+# from ludwig.datasets import titanic
```
Comment (geoffreyangus) [inline examples/titanic/simple_model_training.py:17]: Remove?

File: ludwig/distributed/base.py [hunk 4]
```diff
@@ -20,11 +26,24 @@ class DistributedStrategy(ABC):
...
     @abstractmethod
-    def wrap_model(self, model: nn.Module) -> nn.Module:
+    def prepare(
+        self, model: nn.Module, optimizer: Optimizer, lr_scheduler: "LRScheduler", trainer_config: "ECDTrainerConfig"
+    ) -> Tuple[nn.Module, Optimizer, "LRScheduler"]:
         pass
 
-    @abstractmethod
-    def wrap_optimizer(self, optimizer: Optimizer, model: nn.Module, gradient_accumulation_steps: int) -> Optimizer:
+    def to_device(self, model: nn.Module) -> nn.Module:
```
Comment (geoffreyangus) [inline ludwig/distributed/base.py:34]: should this take device as optional?

File: ludwig/features/base_feature.py [hunk 5]
```diff
@@ -97,6 +98,16 @@ def add_feature_data(
         raise NotImplementedError
 
 
+@dataclass
+class ModuleWrapper:
+    """Used to prevent the PredictModule from showing up an attribute on the feature module.
+
+    This is necessary to avoid inflight errors from DeepSpeed.
```
Comment (geoffreyangus) [inline ludwig/features/base_feature.py:105]: What is an inflight error?

File: ludwig/features/base_feature.py [hunk 6]
```diff
@@ -110,6 +121,9 @@ def __init__(self):
         self.probabilities_key = PROBABILITIES
         self.logits_key = LOGITS
 
+    def forward(self, *args, **kwargs):
```
Comment (geoffreyangus) [inline ludwig/features/base_feature.py:124]: How is this not infinitely recursive?

File: ludwig/models/predictor.py [hunk 7]
```diff
@@ -65,23 +66,33 @@ class Predictor(BasePredictor):
...
-        self.device = get_torch_device() if not model.type() == MODEL_GBM else "cpu"
-        self.model = model.to(self.device)
+        device = get_torch_device()
+        if model.type() == MODEL_GBM:
+            # TODO (jeffkinnison): revert to using the requested device for GBMs when device usage is fixed
+            device = "cpu"
+            dist_model = dist_model.to(device)
+
+        self.device = device
+        self.dist_model = dist_model
```
Comment (geoffreyangus) [inline ludwig/models/predictor.py:90]: Can you leave a comment on the difference between these two models? Maybe as part of an overall class docstring

File: ludwig/backend/deepspeed.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+import deepspeed
+
+from ludwig.backend.base import DataParallelBackend
+from ludwig.constants import FALLBACK_BATCH_SIZE
+from ludwig.distributed import init_dist_strategy
+from ludwig.utils.batch_size_tuner import BatchSizeEvaluator
+
+
+class DeepSpeedBackend(DataParallelBackend):
```
Comment (geoffreyangus) [inline ludwig/backend/deepspeed.py:11]: How does this new backend interact with Ray/Ray Datasets? Noticed that DataParallelBackend uses the LocalPreprocessingMixin, so curious what you're thinking here.

File: ludwig/utils/state_dict_backward_compatibility.py [hunk 9]
```diff
@@ -24,7 +24,19 @@ def _update_transformers_to_freeze_module(state_dict):
...
+def _update_combiner_no_device_tensor(state_dict):
+    """Removed device_tensor from state_dict following DeepSpeed integration."""
+    return {k: v for k, v in state_dict.items() if not k.endswith("device_tensor")}
+
+
 def update_state_dict(state_dict):
     """Checks state_dict on load, updates state dict if needed."""
     state_dict = _update_transformers_to_freeze_module(state_dict)
+    state_dict = _update_combiner_no_input_features(state_dict)
+    state_dict = _update_combiner_no_device_tensor(state_dict)
```
Comment (geoffreyangus) [inline ludwig/utils/state_dict_backward_compatibility.py:41]: Nice change

**Review by geoffreyangus — COMMENTED** [review]
Nice work on this! Left some questions, but overall looking really good.

File: ludwig/backend/deepspeed.py (hunk 8 above)
Comment (tgaddair) [reply ludwig/backend/deepspeed.py:11]: It works the same as the existing HorovodBackend. You run with the command line tool `deepspeed`. It processes the data in-memory (like local training). It doesn't use Ray (that's the DeepSpeedStrategy).

File: ludwig/features/base_feature.py (hunk 6 above)
Comment (tgaddair) [reply ludwig/features/base_feature.py:124]: Hmm, good question! Shouldn't be needed now that we made it an nn.Module again. Removed!

File: examples/llm_finetuning/README.md
Comment (arnavgarg1) [inline examples/llm_finetuning/README.md:1]: Assuming you're going to land instructions to run the configs in the README?

File: examples/llm_finetuning/train_imdb_api.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+# }
+# backend = {
+#     "type": "deepspeed",
+#     "zero_optimization": {"stage": 3},
+# }
+
+# print(model_dir)
+# model = LudwigModel.load(model_dir, backend=backend)
+# results, _ = model.predict("/home/ray/titanic.csv", skip_save_predictions=False)
+# print(results)
```
Comment (arnavgarg1) [inline examples/llm_finetuning/train_imdb_api.py:98]: Cleanup?

Comment (tgaddair) [reply examples/llm_finetuning/train_imdb_api.py:98]: Yeah, this is all messy as I'm still benchmarking and testing some things.

File: ludwig/data/dataset/ray.py [hunk 11]
```diff
@@ -375,7 +375,7 @@ def _fetch_next_batch(self):
             self._next_batch = next(self.dataset_batch_iter)
             # If the batch has only one row and self.ignore_last, skip the batch
             # to prevent batchnorm / dropout related Torch errors
-            if self.ignore_last and len(self._next_batch[self._sample_feature_name]) == 1:
+            if self.batch_size > 1 and self.ignore_last and len(self._next_batch[self._sample_feature_name]) == 1:
```
Comment (arnavgarg1) [inline ludwig/data/dataset/ray.py:378]: nice catch! i noticed this with my prompt tuning PR as well that we go into an infinite loop with batch size 1

File: ludwig/distributed/base.py [hunk 12]
```diff
@@ -20,11 +27,27 @@ class DistributedStrategy(ABC):
     """
 
     @abstractmethod
-    def wrap_model(self, model: nn.Module) -> nn.Module:
+    def prepare(
+        self,
+        model: nn.Module,
+        trainer_config: "ECDTrainerConfig",
+        base_learning_rate: float,
+    ) -> Tuple[nn.Module, Optimizer]:
```
Comment (arnavgarg1) [inline ludwig/distributed/base.py:35]: It might be useful to add a docstring here indicating what the `prepare` function is supposed to do. From below, it seems like we're returning the model/wrapping the model with the relevant strategy initialization + creating the optimizer?

File: ludwig/backend/deepspeed.py (hunk 8 above)
Comment (geoffreyangus) [reply ludwig/backend/deepspeed.py:11]: Nice, just read the PR description. Thanks!

**Review by geoffreyangus — APPROVED** [review]

File: ludwig/distributed/deepspeed.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+)
+
+
+class DeepSpeedStrategy(DDPStrategy):
+    def __init__(self, zero_optimization: Optional[Dict[str, Any]] = None, **kwargs):
+        # If we're initializing from a `deepspeed` CLI command, deepspeed will have already been initialized, as
+        # indicated by the presence of the LOCAL_RANK var. Otherwise, we're initializing from Ray / torchrun, and will
+        # need to set this var ourselves, then init DeepSpeed here.
+        local_rank, local_size = os.environ.get("LOCAL_RANK"), os.environ.get("LOCAL_SIZE")
+        init_deepspeed = local_rank is None or local_size is None
```
Comment (arnavgarg1) [inline ludwig/distributed/deepspeed.py:43]: Do we need both to be set for `deepspeed.init_distributed()` to work?

Comment (arnavgarg1) [inline ludwig/distributed/deepspeed.py:43]: If both of these are set, is the reason we don't need to call `deepspeed.init_distributed()` because deepspeed is already initialized? When would that be the case?

File: ludwig/distributed/deepspeed.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        logging.info("Using DeepSpeed strategy")
+
+    def prepare(
+        self,
+        model: nn.Module,
+        trainer_config: "ECDTrainerConfig",
+        base_learning_rate: float,
+    ) -> Tuple[nn.Module, Optimizer]:
+        # If `batch_size=auto`, we set to 2 temporarily until auto-tuning adjusts it`
+        batch_size = trainer_config.batch_size if isinstance(trainer_config.batch_size, int) else 2
```
Comment (arnavgarg1) [inline ludwig/distributed/deepspeed.py:63]: Should we log for visibility that we are setting the batch_size to 2 if it was `auto`?

File: ludwig/modules/optimization_modules.py [hunk 15]
```diff
@@ -13,50 +13,57 @@
...
-from ludwig.schema.optimizers import BaseOptimizerConfig, GradientClippingConfig, optimizer_registry
 from ludwig.utils.misc_utils import get_from_registry
 from ludwig.utils.torch_utils import LudwigModule
 
+if TYPE_CHECKING:
+    from ludwig.schema.optimizers import BaseOptimizerConfig, GradientClippingConfig
+
+
+def create_clipper(gradient_clipping_config: Optional["GradientClippingConfig"]):
+    from ludwig.schema.optimizers import GradientClippingConfig
```
Comment (arnavgarg1) [inline ludwig/modules/optimization_modules.py:28]: Do we need to import again if it's already imported under TYPE_CHECKING?

File: ludwig/trainers/trainer.py [hunk 16]
```diff
@@ -386,15 +388,26 @@ def tune_batch_size(
...
+                return self.distributed.broadcast_object(best_batch_size)
+            finally:
+                # Restore original parameters to defaults
+                self.skip_save_model = skip_save_model
+                self.skip_save_progress = skip_save_progress
+                self.skip_save_log = skip_save_log
+                if snapshot_weights:
+                    if self.distributed.prepare_before_load():
+                        self.prepare()
+                    self.resume_weights_and_optimizer(str(tmpdir), checkpoint)
```
Comment (arnavgarg1) [inline ludwig/trainers/trainer.py:410]: Can we add a comment for why we're doing this / when we have to do this?

File: ludwig/modules/optimization_modules.py (hunk 15 above)
Comment (tgaddair) [reply ludwig/modules/optimization_modules.py:28]: `TYPE_CHECKING` doesn't import it at all, it's only used for static analysis of the type hints.

Comment (arnavgarg1) [reply ludwig/modules/optimization_modules.py:28]: Ah sorry, my bad

File: ludwig/distributed/deepspeed.py (hunk 13 above)
Comment (tgaddair) [reply ludwig/distributed/deepspeed.py:43]: We don't need to `init_distributed` when using Ray because we're already in a `torch.distributed` environment, but when using the `deepspeed` launch we need to call `init_distributed` to be in that context.

File: ludwig/data/dataset/ray.py (hunk 11 above)
Comment (tgaddair) [reply ludwig/data/dataset/ray.py:378]: Yeah, this was casuing me problems today!

File: examples/llm_finetuning/README.md
Comment (tgaddair) [reply examples/llm_finetuning/README.md:1]: Yep, done.

File: examples/llm_finetuning/train_imdb_api.py (hunk 10 above)
Comment (tgaddair) [reply examples/llm_finetuning/train_imdb_api.py]: Updated.

File: ludwig/distributed/deepspeed.py (hunk 14 above)
Comment (tgaddair) [reply ludwig/distributed/deepspeed.py:63]: It shouldn't be needed because we override it, it's just so DeepSpeed doesn't comnplain.

File: ludwig/trainers/trainer.py [hunk 17]
```diff
@@ -540,23 +555,36 @@ def run_evaluation(
...
+                checkpoint_manager,
             )
         else:
             # There's no validation, so we save the model.
-            if self.is_coordinator() and not self.skip_save_model:
-                self.model.save(save_path)
+            if not self.skip_save_model:
+                logger.info("Saving model.\n")
+                checkpoint_manager.save_best(progress_tracker.steps)
+                self.callback(lambda c: c.on_save_best_checkpoint(self, progress_tracker, save_path))
```
Comment (arnavgarg1) [inline ludwig/trainers/trainer.py:565]: Does this save on the coordinator automatically?

File: ludwig/trainers/trainer.py [hunk 18]
```diff
@@ -765,19 +801,28 @@ def train(self, training_set, validation_set=None, test_set=None, save_path="mod
...
+        if not self.skip_save_model:
+            state_dict = checkpoint_manager.get_best_checkpoint_state_for_inference(self.return_device)
+            if not return_state_dict:
+                if self.distributed.is_model_parallel():
+                    # Assume the full weights cannot fit in memory on GPU
+                    self.model = self.model.cpu()
+                self.model.load_state_dict(state_dict)
+        elif return_state_dict:
+            state_dict = self.model.cpu().state_dict()
+        return_value = self.model if not return_state_dict else state_dict
```
Comment (arnavgarg1) [inline ludwig/trainers/trainer.py:818]: Are you able to leave a comment on why/when we'd want to return the model state dict as opposed to the model with the loaded state dict?

Comment (arnavgarg1) [reply ludwig/trainers/trainer.py:818]: I assume it is because each of the workers can return their individual state dicts (because of the model parallelism), and then it's all combined back on the coordinator, but I can't seem to see the conditional here for is_coordinator() to load the combined state dicts back into the model object

File: ludwig/trainers/trainer.py (hunk 17 above)
Comment (arnavgarg1) [reply ludwig/trainers/trainer.py:565]: Ohhhh I see, they're all CoordinatorCheckpoint classes that internally do the `is_coordinator()` checks, so this makes sense!

File: ludwig/utils/checkpoint_utils.py [hunk 19]
```diff
@@ -65,28 +71,65 @@ def get_latest_checkpoint_path(directory: str) -> str:
...
+
+
+@DeveloperAPI
+class CoordinatorCheckpoint(Checkpoint):
+    def prepare(self, directory: str):
+        if self.distributed.is_coordinator():
+            super().prepare(directory)
+
+    def load(self, save_path: str, device: Optional[torch.device] = None) -> bool:
+        """Load state from a saved checkpoint.
```
Comment (arnavgarg1) [inline ludwig/utils/checkpoint_utils.py:124]: With model parallelism, how does the model get loaded back? Is the idea that we load on the main coordinator device, then shard across devices again? If yes, is there a need to wrap the load call with `self.distributed.is_coordinator()`?

File: ludwig/trainers/trainer.py (hunk 18 above)
Comment (tgaddair) [reply ludwig/trainers/trainer.py:818]: Ray sends the state dict back to the caller, not the model, so loading the weights back into the model is wasteful.

File: ludwig/utils/checkpoint_utils.py (hunk 19 above)
Comment (tgaddair) [reply ludwig/utils/checkpoint_utils.py:124]: Each worker loads is part of the model independently, so each worker needs to call it.

File: ludwig/trainers/trainer.py (hunk 17 above)
Comment (tgaddair) [reply ludwig/trainers/trainer.py:565]: Good point, though, for hyperopt we only want to save the model on the coordinator. Updated.

File: ludwig/distributed/deepspeed.py (hunk 13 above)
Comment (arnavgarg1) [reply ludwig/distributed/deepspeed.py:44]: Ahhh got it, nice!

**Review by arnavgarg1 — APPROVED** [review]
Amazing work, super excited to try deepspeed 🚢  🚀 . Thanks for leaving comments - they definitely make it much more clear!
